File: server/src/main/java/com/im/server/common/PacketCodec.java
package com.im.server.common;

import io.vertx.core.buffer.Buffer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class PacketCodec {
    private static final Logger log = LoggerFactory.getLogger(PacketCodec.class);

    public static ImPacket decode(Buffer buf) {
        if (buf.length() < Cmd.HEADER_LEN) return null;

        int offset = 0;
        int magic = buf.getShort(offset) & 0xFFFF; offset += 2;
        if (magic != Cmd.MAGIC) {
            log.warn("invalid magic: 0x{}", Integer.toHexString(magic));
            return null;
        }

        offset += 2; // version
        int cmd = buf.getShort(offset) & 0xFFFF; offset += 2;
        int msgType = buf.getByte(offset) & 0xFF; offset += 1;
        int sequenceId = buf.getInt(offset); offset += 4;
        int dataLength = buf.getInt(offset); offset += 4;
        offset += 7; // padding

        if (buf.length() < Cmd.HEADER_LEN + dataLength) return null;

        byte[] bodyBytes = buf.getBytes(offset, offset + dataLength);
        return new ImPacket(cmd, msgType, sequenceId, bodyBytes);
    }

    public static Buffer encode(ImPacket packet) {
        return packet.encode();
    }
}

File: client-android/app/src/main/java/com/im/client/data/remote/PacketCodec.kt
package com.im.client.data.remote

import java.nio.ByteBuffer

object PacketCodec {

    fun decode(buf: ByteArray, offset: Int, length: Int): ImPacket? {
        if (length < Cmd.HEADER_LEN) return null

        val bb = ByteBuffer.wrap(buf, offset, length)
        val magic = bb.short.toInt() and 0xFFFF
        if (magic != Cmd.MAGIC) return null

        bb.short // skip version
        val cmd = bb.short.toInt() and 0xFFFF
        val msgType = bb.get().toInt() and 0xFF
        val sequenceId = bb.int
        val dataLength = bb.int
        bb.get(ByteArray(7)) // skip padding

        if (length < Cmd.HEADER_LEN + dataLength) return null

        val bodyBytes = ByteArray(dataLength)
        bb.get(bodyBytes)
        return ImPacket(cmd, msgType, sequenceId, bodyBytes)
    }

    fun decode(buf: ByteArray): ImPacket? = decode(buf, 0, buf.size)
}
